### ISSUE TITLE
Render image from KaTeX

### ISSUE THREAD

**Issue by lukeburns** [Khan/KaTeX#328]
Is there any way I can render a png of arbitrary size from KaTeX? I've tried taking screenshots using PhantomJS, but it doesn't give me great control over the image (cropping is wonky).

**Comment by xymostech** [comment]
Seeing as you're spawning a new process of something anyway, is there a reason you're not just using normal TeX/LaTeX to render your math?

**Comment by lukeburns** [comment]
That's an option. Seems like it would be more processor intensive since it requires compiling TeX to a pdf then converting to png.

I'm using KaTeX for preview of the image, and ideally the png should match up with the preview.

**Comment by eugene-eeo** [comment]
@lukeburns you can check out #251 for rendering to canvas. I'm also very interested in this, but it seems that currently the direction, AFAIK is heading towards refactoring the core and trying to keep it small and performant, so it seems that we'll have to wait a bit longer for this feature to be implemented "properly". Currently I am doing it with [html2canvas](https://github.com/niklasvh/html2canvas) and trimming the canvas to get rid of the whitespace. This is a bit of a hack but it's the best currently.

**Comment by Johnz86** [comment]
I have problem with rendering of math equation in katex to image from canvas. Currently my result looks like this:
![image](https://user-images.githubusercontent.com/2708520/68480403-00503480-0235-11ea-8b62-d0c4c5b95b16.png)
Currently all solutions that I came across use canvas. They all brake after rendering more complex equations. They fail to display proper shapes and to allign equation elements. The example image was produced by ```html2canvas``` and similarly broken results are with ```rasterizeHTML``` and others.
Since the last post in this thread was in 2015, what are current ways on client side to render to image?

**Comment by kevinbarabash** [comment]
You could look at how we do the screenshotting tests (it uses docker and real browsers).  It's possible a little overkill.  You may also want to look at #504 and #251.

**Comment by csbenjamin** [comment]
@Johnz86 Were you able to fix those issues?

**Comment by Johnz86** [comment]
> @Johnz86 Were you able to fix those issues?
I regret to admit, that I went with mathjax. I had limited time for finding a solution.
Its half a year since then. Did you try with newest version of this lib?

**Comment by csbenjamin** [comment]
@Johnz86 Yes, I tried with the newest version of this lib and html2canvas. I will try mathjax. Although I'm also thinking to try using headless chrome on the server to take screenshot. On the one hand, mathjax seams to be slower, on the other hand, using headless chrome on the server consumes server resources. I'm going to think what is the best solution for me. Thanks for you attention.

**Comment by mbourne** [comment]
@csbenjamin I'm not sure what your application involves, nor what your server OS is, but you may wish to consider the approach I took for my "Send math in emails" page:

https://www.intmath.com/help/send-math-email.php

It uses KaTeX for initial user editing, then on submit, it uses Tex2Png to convert the math to images: https://github.com/Gregwar/Tex2png

Notes: 
1. I make use of ASCIIMath for math entry on that page, and this converts it to LaTeX ready for KaTeX. If you don't need that step, then it's even easier.

2. Tex2png produces a nice crisp image, but Cloudflare in its wisdom does an image optimize step, so the outcome is not a clear as I would like. No Cloudflare and the images are much better.

**Comment by csbenjamin** [comment]
@mbourne thanks for your message. My use case is the following. I have an app that gives a whiteboard to the user. The user can draw on a canvas. Also, que user can add stick notes to the board and these notes accept Tex. The whiteboard has a button for downloading the board. So, I need to convert all elements on the page to png. The elements include the canvas where the user draws and div elements with katex. Given the complexity of my case, I went for using selenium and headless chrome on the server and taking screenshot. That way, the image will be exactly what the user is seeing.

**Comment by sdos-lehr** [comment]
I have found an answer on SO to the problem shown in the previously posted image 

> I have problem with rendering of math equation in katex to image from canvas. Currently my result looks like this:
> ![image](https://user-images.githubusercontent.com/2708520/68480403-00503480-0235-11ea-8b62-d0c4c5b95b16.png)
> Currently all solutions that I came across use canvas. They all brake after rendering more complex equations. They fail to display proper shapes and to allign equation elements. The example image was produced by `html2canvas` and similarly broken results are with `rasterizeHTML` and others.
> Since the last post in this thread was in 2015, what are current ways on client side to render to image?

[https://stackoverflow.com/a/59162619/6330074](url) suggested that the problem is a conflict between using css width and svg width attributes and adding the following code snippet before conversion helped in my case:
`var svgElements = document.body.querySelectorAll('svg');
svgElements.forEach(function(item) {
    item.setAttribute("width", item.getBoundingClientRect().width);
    item.style.width = null;
});`

**Comment by csbenjamin** [comment]
@sdos-lehr The problem is, katex does not output svg. Your solution is for the problem where svg element is not shown. That is not the case here.

**Comment by crystalYY** [comment]
> I have found an answer on SO to the problem shown in the previously posted image
> 
> > I have problem with rendering of math equation in katex to image from canvas. Currently my result looks like this:
> > ![image](https://user-images.githubusercontent.com/2708520/68480403-00503480-0235-11ea-8b62-d0c4c5b95b16.png)
> > Currently all solutions that I came across use canvas. They all brake after rendering more complex equations. They fail to display proper shapes and to allign equation elements. The example image was produced by `html2canvas` and similarly broken results are with `rasterizeHTML` and others.
> > Since the last post in this thread was in 2015, what are current ways on client side to render to image?
> 
> [https://stackoverflow.com/a/59162619/6330074](url) suggested that the problem is a conflict between using css width and svg width attributes and adding the following code snippet before conversion helped in my case:
> `var svgElements = document.body.querySelectorAll('svg'); svgElements.forEach(function(item) { item.setAttribute("width", item.getBoundingClientRect().width); item.style.width = null; });`

that's work for me!! thanks
